test/e2e/page-objects/benchmark/page-load-benchmark.ts

```diff
@@ -0,0 +1,471 @@
+   *
+   * @returns Array of statistical summaries, one for each tested page
+   */
+  calculateStatistics(): BenchmarkSummary[] {
+    const summaries: BenchmarkSummary[] = [];
+
+    const resultsByPage = this.results.reduce(
+      (acc, result) => {
+        if (!acc[result.page]) {
+          acc[result.page] = [];
+        }
+        acc[result.page].push(result);
+        return acc;
+      },
+      {} as Record<string, BenchmarkResult[]>,
+    );
+
+    for (const [page, pageResults] of Object.entries(resultsByPage)) {
+      const metrics = pageResults.map((r) => r.metrics);
+      const metricKeys = Object.keys(metrics[0]) as (keyof Omit<
+        BenchmarkMetrics,
+        'memoryUsage'
+      >)[];
+
+      const summary: BenchmarkSummary = {
+        page,
+        samples: metrics.length,
+        mean: {},
+        p95: {},
+        p99: {},
+        min: {},
+        max: {},
+        standardDeviation: {},
+      };
+
+      for (const key of metricKeys) {
+        const values = metrics
+          .map((m) => m[key])
+          .filter((v) => typeof v === 'number') as number[];
+        if (values.length > 0) {
+          summary.mean[key] = this.calculateMean(values);
+          summary.p95[key] = this.calculatePercentile(values, 95);
+          summary.p99[key] = this.calculatePercentile(values, 99);
+          summary.min[key] = Math.min(...values);
+          summary.max[key] = Math.max(...values);
+          summary.standardDeviation[key] =
+            this.calculateStandardDeviation(values);
+        }
+      }
+
+      summaries.push(summary);
+    }
+
+    return summaries;
+  }
+
+  /**
+   * Calculates the arithmetic mean (average) of an array of numbers.
+   *
+   * @param values - Array of numeric values
+   * @returns The mean value
+   */
+  private calculateMean(values: number[]): number {
+    return values.reduce((sum, val) => sum + val, 0) / values.length;
+  }
+
+  /**
+   * Calculates the specified percentile value from an array of numbers.
+   * Uses the nearest-rank method for percentile calculation.
+   *
+   * @param values - Array of numeric values
+   * @param percentile - Percentile to calculate (0-100)
+   * @returns The value at the specified percentile
+   */
+  private calculatePercentile(values: number[], percentile: number): number {
+    const sorted = [...values].sort((a, b) => a - b);
+    const index = Math.floor((percentile / 100) * sorted.length);
+    return sorted[index];
+  }
+
+  /**
+   * Calculates the standard deviation of an array of numbers.
+   * Uses the population standard deviation formula.
+   *
+   * @param values - Array of numeric values
+   * @returns The standard deviation value
+   */
+  private calculateStandardDeviation(values: number[]): number {
+    const mean = this.calculateMean(values);
+    const squaredDiffs = values.map((val) => Math.pow(val - mean, 2));
+    const variance = this.calculateMean(squaredDiffs);
+    return Math.sqrt(variance);
+  }
```

Comment:
writing your own basic functions and not importing them from a package? Is this allowed in javascript? I'm calling the npm police